src/java.base/share/classes/java/io/Bits.java

```diff
@@ -41,39 +50,31 @@ static boolean getBoolean(byte[] b, int off) {
     }
 
     static char getChar(byte[] b, int off) {
-        return (char) ((b[off + 1] & 0xFF) +
-                       (b[off] << 8));
+        return (char) (short) SHORT.get(b, off);
     }
 
     static short getShort(byte[] b, int off) {
-        return (short) ((b[off + 1] & 0xFF) +
-                        (b[off] << 8));
+        return (short) SHORT.get(b, off);
     }
 
     static int getInt(byte[] b, int off) {
-        return ((b[off + 3] & 0xFF)      ) +
-               ((b[off + 2] & 0xFF) <<  8) +
-               ((b[off + 1] & 0xFF) << 16) +
-               ((b[off    ]       ) << 24);
+        return (int) INT.get(b, off);
     }
 
     static float getFloat(byte[] b, int off) {
-        return Float.intBitsToFloat(getInt(b, off));
+        // Using Float.intBitsToFloat collapses NaN values to a single
+        // "canonical" NaN value
+        return Float.intBitsToFloat((int) INT.get(b, off));
     }
 
     static long getLong(byte[] b, int off) {
-        return ((b[off + 7] & 0xFFL)      ) +
-               ((b[off + 6] & 0xFFL) <<  8) +
-               ((b[off + 5] & 0xFFL) << 16) +
-               ((b[off + 4] & 0xFFL) << 24) +
-               ((b[off + 3] & 0xFFL) << 32) +
-               ((b[off + 2] & 0xFFL) << 40) +
-               ((b[off + 1] & 0xFFL) << 48) +
-               (((long) b[off])      << 56);
+        return (long) LONG.get(b, off);
     }
 
     static double getDouble(byte[] b, int off) {
-        return Double.longBitsToDouble(getLong(b, off));
+        // Using Double.longBitsToDouble collapses NaN values to a single
+        // "canonical" NaN value
+        return Double.longBitsToDouble((long) LONG.get(b, off));
```

Comment:
For the records, I believe the template classes for byte array var handle already perform this conversion - e.g.

```
@ForceInline
        static $type$ get(VarHandle ob, Object oba, int index) {
            ArrayHandle handle = (ArrayHandle)ob;
            byte[] ba = (byte[]) oba;
#if[floatingPoint]
            $rawType$ rawValue = UNSAFE.get$RawType$Unaligned(
                    ba,
                    ((long) index(ba, index)) + Unsafe.ARRAY_BYTE_BASE_OFFSET,
                    handle.be);
            return $Type$.$rawType$BitsTo$Type$(rawValue); // <-----------------------------
#else[floatingPoint]
            return UNSAFE.get$Type$Unaligned(
                    ba,
                    ((long) index(ba, index)) + Unsafe.ARRAY_BYTE_BASE_OFFSET,
                    handle.be);
#end[floatingPoint]
        }
 ````

Reply:
I see - thanks for the explanation - the problem is with `floatToIntBits` vs `floatToRawIntBits` in the "put" operation. The get operation is ok, but then having asymmetry where we use a float VH in one case (get) but not in the other (put) is not desirable.

Reply:
@mcimadamore According to the spec, `floatToRawIntBits` and `doubleToRawLongBits` ensure that the original input bits are all preserved, even for NaN values.
On the other hand, the spec says that `intBitsToFloat` and `longBitsToDouble` only ensure a best effort in preserving the output bits of NaN values.
So I think `VarHandle` `get` and `put` do their best to preserve the bits, but no more than the underlying platform can possibly offer.
This behavior for floating-point values is not fully specified in `VarHandle`, so there's room for improvement in the doc.